src/core/module.js

```diff
@@ -0,0 +1,146 @@
+const wasmSupported = () => {
+    let supported = null;
+
+    if (supported === null) {
+        supported = (() => {
+            try {
+                if (typeof WebAssembly === "object" && typeof WebAssembly.instantiate === "function") {
+                    const module = new WebAssembly.Module(Uint8Array.of(0x0, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00));
+                    if (module instanceof WebAssembly.Module)
+                        return new WebAssembly.Instance(module) instanceof WebAssembly.Instance;
+                }
+            } catch (e) { }
+            return false;
+        })();
+    }
+
+    return supported;
+};
+
+// load a script
+const loadScript = (url, callback) => {
+    const s = document.createElement('script');
+    s.setAttribute('src', url);
+    s.onload = () => {
+        callback(null);
+    };
+    s.onerror = () => {
+        callback(`Failed to load script='${url}'`);
+    };
+    document.body.appendChild(s);
+};
+
+// load a wasm module
+const loadWasm = (moduleName, config, callback) => {
+    const loadUrl = (wasmSupported() && config.glueUrl && config.wasmUrl) ? config.glueUrl : config.fallbackUrl;
+    if (loadUrl) {
+        loadScript(loadUrl, (err) => {
+            if (err) {
+                callback(err, null);
+            } else {
+                window[moduleName]({
+                    locateFile: () => config.wasmUrl
+                }).then((instance) => {
+                    callback(null, instance);
+                });
+            }
+        });
+    } else {
+        callback('No supported wasm modules found.', null);
+    }
+};
+
+// get state object for the named module
+const getModule = (() => {
+    const modules = {};
+    return (name) => {
+        if (!modules.hasOwnProperty(name)) {
+            modules[name] = {
+                config: null,
+                initializing: false,
+                instance: null,
+                error: null,
+                callbacks: []
+            };
+        }
+        return modules[name];
+    };
+})();
+
+const initialize = (moduleName, module) => {
+    if (module.initializing) {
+        return;
+    }
+
+    const config = module.config;
+
+    if (config.glueUrl || config.wasmUrl || config.fallbackUrl) {
+        module.initializing = true;
+        loadWasm(moduleName, config, (err, instance) => {
+            if (err) {
+                console.error(`failed to initialize module=${moduleName} error=${err}`);
+            }
+            module.error = err;
+            module.instance = instance;
+            module.callbacks.forEach((callback) => {
+                callback(err, instance);
+            });
+        });
+    }
+};
+
+/**
+ * Callback used by {@link Module#getInstance}.
+ *
+ * @callback ModuleInstanceCallback
+ * @param {string} error - If the instance fails to load this will contain a description of the error.
+ * @param {any} moduleInstance - The module instance.
+ */
+
+class Module {
```

Comment:
The `Module` names fees little too generic here.

Reply:
I thought about it, but `EMSModule` didn't seem great either.

Reply:
WasmModule? Could a module be something else that isn't Wasm?

Reply:
It's a subset of wasm modules - those generated by emscripten compiler.

Reply:
renamed to `WasmModule`